### ISSUE TITLE
[Newbie Thoughts] Username fails validation

### ISSUE THREAD

**Issue by chrisjlocke** [sqlitebrowser/dbhub.io#82]
Creating a wacky username fails validation, but the user doesn't know why.

![image](https://user-images.githubusercontent.com/7872218/29310926-9cae2f6e-81a7-11e7-8e3d-9130edbfb984.png)

This one could be a nice simple one for me to play with as the code to fail is already there - just needs a 'heres why, knobby' text added.  Maybe.... ?

**Comment by justinclift** [comment]
Yep, that'd definitely be useful. :smile:

The code which does that is this:

* Server side
  https://github.com/sqlitebrowser/dbhub.io/blob/37b2e668bc5a68cef4ed833e3fdc15d7b3c13bac/webui/main.go#L671-L703

* Browser side
  https://github.com/sqlitebrowser/dbhub.io/blob/37b2e668bc5a68cef4ed833e3fdc15d7b3c13bac/webui/templates/selectusername.html#L57-L65

It might seem (probably deceptively) simple. :wink:

The concept of is obviously straight forward:

1. the javascript on the front end sends a potential username to the server
2. the server runs a few checks, then returns 'y' or 'n' to indicate success or failure
3. the javascript then figures out what to show

The 'y' or 'n' thing is because when writing that code I didn't (at the time) understand how to send http status codes from the server back to the browser.  Thus sending y/n instead.

Too make this a much nicer experience for the user, it'd probably need these bits to be done:

1. Update the server side to send a useful status message + use http status code to indicate success or failure (instead of the y/n thing)
2. Update the browser side to grok and display the useful status message sent
3. (this bit could be more tricky) Adjust the server side validation and/or ReservedUsernamesCheck() check code to return better info, so the above two pieces have something to work with.

On that note, do you have Go 1.9 and Gogland installed? :smile:

**Comment by justinclift** [comment]
... and now I realise you're going to need to set up a running local dbhub.io server.

Probably about time I started writing those long overdue "How to setup and run your own dbhub.io server" docs.  Maybe starting with just the bare essentials first. :smile:

**Comment by justinclift** [comment]
Oh... and on Windows too.  This could be fun.

**Comment by justinclift** [comment]
Looking through our dependencies list, we rely on these bits of software:

* Golang - works on Windows
* Minio - works on Windows
* PostgreSQL - works on Windows
* Memcached - doesn't seem to support Windows.  Might work with the new fangled Win 10 Linux/bash layer though? (no idea)

You might need to run Memcached in a VM of some sort if it turns out to really not like Windows.
Alternatively, I can pretty easily setup a new Memcached VM online and update its firewall so only you can connect to it.  We literally have a working memcached server image ready to go in Rackspace that I'd just need to spin up and change some IP info for. :smile:

**Comment by justinclift** [comment]
> The beauty of status codes. I assume (for example) return 450 which means invalid length, 451 which means it contains a + (or invalid characters, but denoting which one is useful) 452 for ...... etc.

Yep.  I'm not an expert on them, more a "lets get it working ok first"-pert.  So... err... get it working "however you can", and we can refine it from there? :smile:

**Comment by chrisjlocke** [comment]
>  I'm ... more a "lets get it working ok first"-pert.

![image](https://user-images.githubusercontent.com/7872218/29323142-e1a9519a-81d7-11e7-9f31-313a5c1a4a34.png)
